Duplicati/Library/Main/Operation/DeleteHandler.cs

```diff
@@ -221,34 +228,41 @@ private DateTime[] GetFilesetsToDelete(Database.LocalDeleteDatabase db, DateTime
             var fullVersionsKeptCount = 0;
             if (fullVersionsToKeep > 0 && fullVersionsToKeep < backupsRemaining.Count)
             {
-                // keep the number of full backups specified in fullVersionsToKeep.
-                // once the last full backup t okeep is found, also keep the partials immediately after it the full backup.
-                // add the remainder of full and partial backups to toDelete
-                bool foundLastFullBackupToKeep = false;
+                ISet<DateTime> intermediatePartials = new HashSet<DateTime>();
+                bool haveFullBackup = false;
+
+                // Keep the number of full backups specified in fullVersionsToKeep.
+                // Remove partial backups that are followed by full backups.
+                // Once enough versions are kept, delete all older backups.
                 foreach (var backup in backupsRemaining)
                 {
-                    bool isFullBackup;
-                    if (fullVersionsKeptCount < fullVersionsToKeep)
+                    bool isFullBackup = db.IsFilesetFullBackup(backup);
+                    if (isFullBackup)
                     {
-                        isFullBackup = db.IsFilesetFullBackup(backup);
-                        // count only a full backup
-                        if (fullVersionsKeptCount < fullVersionsToKeep && isFullBackup)
+                        toDelete.AddRange(intermediatePartials);

```

Comment:
Maybe there is a fancier way to do the iterating in a few lines like you did with keepTime, using linq. Like get the last n or n-1 full backups and throw away the rest, plus some pending partial ones.

Reply:
I tried several different approaches and ended up with the one in revision 20eb785f098dfa5315f270687a6824b4374748ee.  Let me know if it looks correct and if you think it is more readable.